src/View/SSViewer_DataPresenter.php

```diff
@@ -40,95 +40,126 @@ class SSViewer_DataPresenter extends SSViewer_Scope
 
     /**
      * Underlay variables. Concede precedence to overlay variables or anything from the current scope
-     * @var array|null
+     *
+     * @var array
      */
     protected $underlay;
 
-    public function __construct($item, $overlay = null, $underlay = null, $inheritedScope = null)
-    {
+    /**
+     * @var object $item
+     * @var array $overlay
+     * @var array $underlay
+     * @var SSViewer_Scope $inheritedScope
+     */
+    public function __construct(
+        $item,
+        array $overlay = null,
+        array $underlay = null,
+        SSViewer_Scope $inheritedScope = null
+    ) {
         parent::__construct($item, $inheritedScope);
 
-        // Build up global property providers array only once per request
-        if (self::$globalProperties === null) {
-            self::$globalProperties = array();
-            // Get all the exposed variables from all classes that implement the TemplateGlobalProvider interface
-            $this->createCallableArray(
-                self::$globalProperties,
-                TemplateGlobalProvider::class,
-                "get_template_global_variables"
-            );
+        $this->overlay = $overlay ?: [];
+        $this->underlay = $underlay ?: [];
+
+        $this->cacheGlobalProperties();
+        $this->cacheIteratorProperties();
+    }
+
+    /**
+     * Build cache of global properties
+     */
+    protected function cacheGlobalProperties()
+    {
+        if (self::$globalProperties !== null) {
+            return;
         }
 
-        // Build up iterator property providers array only once per request
-        if (self::$iteratorProperties === null) {
-            self::$iteratorProperties = array();
-            // Get all the exposed variables from all classes that implement the TemplateIteratorProvider interface
-            // //call non-statically
-            $this->createCallableArray(
-                self::$iteratorProperties,
-                TemplateIteratorProvider::class,
-                "get_template_iterator_variables",
-                true
-            );
+        self::$globalProperties = $this->getPropertiesFromProvider(
+            TemplateGlobalProvider::class,
+            'get_template_global_variables'
+        );
+    }
+
+    /**
+     * Build cache of global iterator properties
+     */
+    protected function cacheIteratorProperties()
+    {
+        if (self::$iteratorProperties !== null) {
+            return;
         }
 
-        $this->overlay = $overlay ? $overlay : array();
-        $this->underlay = $underlay ? $underlay : array();
+        self::$iteratorProperties = $this->getPropertiesFromProvider(
+            TemplateIteratorProvider::class,
+            'get_template_iterator_variables',
+            true // Call non-statically
+        );
     }
 
-    protected function createCallableArray(&$extraArray, $interfaceToQuery, $variableMethod, $createObject = false)
```

Comment:
Method refactored/renamed, not part of our “public” API but happy to undo this if others disagree